mods/default/nodes.lua

```diff
@@ -586,30 +586,121 @@ minetest.register_node("default:snow", {
 	paramtype = "light",
+	paramtype2 = "leveled",
 	buildable_to = true,
 	floodable = true,
+	walkable = true,
 	drawtype = "nodebox",
+	leveled = 7,
 	node_box = {
-		type = "fixed",
-		fixed = {
-			{-0.5, -0.5, -0.5, 0.5, -0.25, 0.5},
-		},
+		type = "leveled",
+		fixed = {-0.5, -0.5, -0.5, 0.5, -0.5, 0.5},
 	},
 	collision_box = {
-		type = "fixed",
-		fixed = {
-			{-0.5, -0.5, -0.5, 0.5, -7 / 16, 0.5},
-		},
+		type = "leveled",
+		fixed = {-0.5, -0.5, -0.5, 0.5, -0.5, 0.5},
+	},
+	selection_box = {
+		type = "leveled",
+		fixed = {-0.5, -0.5, -0.5, 0.5, -0.5, 0.5},
 	},
 	groups = {crumbly = 3, falling_node = 1, snowy = 1},
 	sounds = default.node_sound_snow_defaults(),
 
 	on_construct = function(pos)
-		pos.y = pos.y - 1
-		if minetest.get_node(pos).name == "default:dirt_with_grass" then
-			minetest.set_node(pos, {name = "default:dirt_with_snow"})
+		pos.y = pos.y-1
+		local node = minetest.get_node(pos)
+		if node.name == "default:dirt"
+		or node.name == "default:dirt_with_grass"
+		or node.name == "default:dirt_with_grass_footsteps"
+		or node.name == "default:dirt_with_rainforest_litter"
+		or node.name == "default:dirt_with_coniferous_litter" then
+			node.name = "default:dirt_with_snow"
+			minetest.set_node(pos, node)
 		end
 	end,
+
+	-- Handle node drops due to node level.
+	on_dig = function(pos, node, digger)
+		local level = minetest.get_node_level(pos)
+		minetest.node_dig(pos, node, digger)
+		if minetest.get_node(pos).name ~= node.name then
+			local inv = digger:get_inventory()
+			if not inv then
+				return
+			end
+			local left = inv:add_item("main", "default:snow "..tostring(level/7-1))
+			if not left:is_empty() then
+				minetest.add_item({
+					x = pos.x + math.random()/2-0.25,
+					y = pos.y + math.random()/2-0.25,
+					z = pos.z + math.random()/2-0.25,
+				}, left)
+			end
+		end
+	end,
+
+	-- Manage snow levels.
+	on_place = function(itemstack, player, pt)
+		local oldnode_under = minetest.get_node(pt.under)
+		if not oldnode_under then
+			return itemstack, false
+		end
+
+		local olddef_under = minetest.registered_nodes[oldnode_under.name]
+		if not olddef_under then
+			return itemstack, false
+		end
+
+		-- If node under is buildable_to, place into it instead (eg. snow)
+		local pos, node
+		if olddef_under.buildable_to then
+			pos = pt.under
+			node = oldnode_under
+		else
+			pos = pt.above
+			node = minetest.get_node(pos)
+			local def = minetest.registered_nodes[node.name]
+			if not def
+			or not def.buildable_to then
+				return itemstack, false
+			end
+		end
+
+		if node.name ~= "default:snow" then
+			if minetest.get_node{x=pos.x, y=pos.y-1, z=pos.z}.name ==
+					"default:snow" then
+				-- grow the snow below (fixes levelled problem)
+				pos.y = pos.y - 1
+			else
+				-- place a snow
+				return minetest.item_place_node(itemstack, player, pt)
+			end
+		end
+
+		-- grow the snow
+		local level = minetest.get_node_level(pos)
+		level = level + 7
+		if level < 64 then
```

Comment:
If `level` is 63 that's a full snow block so a "default:snowblock" should be placed.

Reply:
[`lua_api.txt` says](
https://github.com/minetest/minetest/blob/95a37efc31586babadc2f42e310e13e8319158a3/doc/lua_api.txt#L6090):
>The nodebox height is 'leveled' / 64 nodes.

Ergo the level has to be 64 in order to be a full node.

Reply:
Done.

Reply:
Sorry, i got it wrong. Indeed, 8 is a 1/8th slab, 64 is a full node.